src/main/kotlin/com/sourcegraph/cody/listeners/CodyCaretListener.kt

```diff
@@ -0,0 +1,38 @@
+package com.sourcegraph.cody.listeners
+
+import com.intellij.openapi.command.CommandProcessor
+import com.intellij.openapi.editor.event.CaretEvent
+import com.intellij.openapi.editor.event.CaretListener
+import com.sourcegraph.cody.agent.CodyAgent
+import com.sourcegraph.cody.agent.CodyAgentService
+import com.sourcegraph.cody.agent.protocol.ProtocolTextDocument
+import com.sourcegraph.cody.autocomplete.CodyAutocompleteManager
+import com.sourcegraph.cody.vscode.InlineCompletionTriggerKind
+import com.sourcegraph.config.ConfigUtil
+import com.sourcegraph.utils.CodyEditorUtil
+
+class CodyCaretListener : CaretListener {
+  override fun caretPositionChanged(e: CaretEvent) {
+    if (!ConfigUtil.isCodyEnabled()) {
+      return
+    }
+
+    val commandName = CommandProcessor.getInstance().currentCommandName
+    if (commandName == CodyEditorUtil.VIM_EXIT_INSERT_MODE_ACTION) {
+      return
+    }
+
+    val editor = e.editor
+    val project = editor.project
+    val textDocument = ProtocolTextDocument.fromEditor(editor)
+    if (project != null && textDocument != null) {
+      CodyAgentService.withAgent(project) { agent: CodyAgent ->
+        agent.server.textDocumentDidFocus(textDocument)
+      }
+    }
```

Comment:
This is fine, I wonder if we want `CodyAgentService.maybeWithEditorDocument(editor) { ...` that takes `agent` and `textDocument`?

Reply:
I see where you are coming from as I do not like that boilerplate as well.
But I do not like to pollute  `CodyAgentService` too much with various utility functions either.
Let me try to rework this a bit!

Reply:
It should look better now :)